File: rest/src/main/java/my/sample/rest/impl/ItemServiceImpl.java
package my.sample.rest.impl;

import my.sample.activator.SampleLogger;
import my.sample.rest.api.Item;
import my.sample.rest.api.ItemService;

import javax.ws.rs.NotFoundException;
import javax.ws.rs.core.Response;
import java.net.URI;
import java.util.List;

/**
 * @author mohammad shamsi <m.h.shams@gmail.com>
 */
public class ItemServiceImpl implements ItemService {
    private ItemRepository repository;
    private SampleLogger logger;

    public void setLogger(SampleLogger logger) {
        this.logger = logger;
    }

    public void setRepository(ItemRepository repository) {
        this.repository = repository;
    }

    @Override
    public List<Item> findAll() {
        logger.warn("GET:: findAll");
        return repository.get();
    }

    @Override
    public Item find(String id) {
        logger.warn("GET:: find/" + id);
        try {
            return repository.get(id);
        } catch (IllegalArgumentException e) {
            throw new NotFoundException(e.getMessage(), e);
        }
    }

    @Override
    public Response update(String id, Item item) {
        logger.warn("PUT:: update/" + id);
        item.setId(id);

        try {
            Item updated = repository.update(item);
            return Response.ok(updated).build();

        } catch (IllegalArgumentException e) {
            throw new NotFoundException(e.getMessage(), e);
        }
    }

    @Override
    public Response add(Item item) {
        logger.warn("POST:: add/" + item.getId());

        try {
            Item added = repository.add(item);
            return Response.created(URI.create(added.getId())).build();

        } catch (IllegalArgumentException e) {
            return Response.notModified(e.getMessage()).build();
        }
    }

    @Override
    public Response delete(String id) {
        logger.warn("DELETE:: delete/" + id);

        try {
            Item item = repository.delete(id);
            return Response.ok(item).build();

        } catch (IllegalArgumentException e) {
            throw new NotFoundException(e.getMessage(), e);
        }
    }
}

File: rest/src/main/kotlin/my/sample/rest/impl/ItemServiceImpl.kt
package my.sample.rest.impl

import my.sample.activator.SampleLogger
import my.sample.rest.api.Item
import my.sample.rest.api.ItemService

import javax.ws.rs.Produces as produces
import javax.ws.rs.GET as get
import javax.ws.rs.PUT as put
import javax.ws.rs.POST as post
import javax.ws.rs.DELETE as delete
import javax.ws.rs.Path as path
import javax.ws.rs.PathParam as pathParam
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

/**
 * @author mohammad shamsi <m.h.shams@gmail.com>
 */
//produces(array(MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON))
public class ItemServiceImpl : ItemService {
    private val repository = ItemRepository()

    private var logger: SampleLogger? = null;

    get
    path("/")
    override fun findAll(): List<Item> {
        logger?.warn("GET:: findAll")
        return repository.get()
    }

    get
    path("/{id}")
    override fun find(pathParam("id") id: String): Item {
        logger?.warn("GET:: find/$id")
        return repository.get(id)
    }

    put
    path("/{id}")
    override fun update(pathParam("id") id: String, item: Item): Response {
        logger?.warn("PUT:: update/$id")
        item.id = id

        try {
            val updated = repository.update(item)
            return Response.ok(updated)!!.build()!!

        } catch (e: IllegalArgumentException) {
            return Response.notModified(e.getMessage())!!.build()!!
        }
    }

    post
    path("/")
    override fun add(item: Item): Response {
        logger?.warn("POST:: add/${item.id}")

        try {
            val added = repository.add(item)
            return Response.ok(added)!!.build()!!

        } catch (e: IllegalArgumentException) {
            return Response.notModified(e.getMessage())!!.build()!!
        }
    }

    delete
    path("/{id}")
    override fun delete(pathParam("id") id: String): Response {
        logger?.warn("DELETE:: delete/$id")

        try {
            val item = repository.delete(id)
            return Response.ok(item)!!.build()!!

        } catch (e: IllegalArgumentException) {
            return Response.notModified(e.getMessage())!!.build()!!
        }
    }
}